theano/gof/type.py

```diff
@@ -795,3 +796,241 @@ def signature(self):
         return (self.type,)
 
 CDataType.Constant = CDataTypeConstant
+
+
+class EnumType(Type, dict):
+    """
+    Op parameter class that allows to create enumerations of constant values.
+
+     - Constants are available as object attributes in Python code and as macro-defined constants in C code.
+     - Constants can be floating values, integers, or booleans (automatically converted to integers).
+     - Constants name must start with a capital letter and contain capital letters, underscores or digits.
+
+    Example::
+
+        enum = EnumType(CONSTANT_1=1, CONSTANT_2=2.5, CONSTANT_3=False, CONSTANT_4=True)
+        print (enum.CONSTANT_1, enum.CONSTANT_2, enum.CONSTANT_3, enum.CONSTANT_4)
+        # will print 1 2.5 0 1
+
+    In C code:
+
+    .. code-block:: c
+
+        int constant_1 = CONSTANT_1;
+        double constant_2 = CONSTANT_2;
+        int constant_3 = CONSTANT_3; // constant_3 == 0
+        int constant_4 = CONSTANT_4; // constant_4 == 1
+
+    You can also specify a C type for the op param if you want to pass one of these constant values at runtime.
+    Default C type is ``double``.
+
+    .. code-block:: python
+
+        enum = EnumType(CONSTANT_1=0, CONSTANT_2=1, CONSTANT_3=2, ctype='size_t')
+        op_param_value = enum.CONSTANT_1
+
+    In C code:
+
+    .. code-block:: c
+
+        size_t value = op_param_value; // contains enum.CONSTANT_1, i.e 0
+
+    .. note::
+
+        This Type (and subclasses) is not complete and should never be used for regular graph operations.
+
+    """
+
+    def check_ctype(self):
+        # C type may be a list of keywords, e.g. "unsigned long long".
+        # We should check each part.
+        if not all(re.match('^[A-Za-z_][A-Za-z0-9_]*$', el) for el in self.ctype.split()):
+            raise TypeError('%s: invalid C type' % type(self).__name__)
+
+    def __init__(self, **kwargs):
+        self.ctype = kwargs.pop('ctype', 'double')
+        self.check_ctype()
+        for k in kwargs:
+            if re.match('^[A-Z][A-Z0-9_]*$', k) is None:
+                raise AttributeError('%s: invalid enum name: "%s". '
+                                     'Only capital letters, underscores and digits '
+                                     'are allowed.' % (type(self).__name__, k))
+            if isinstance(kwargs[k], bool):
+                kwargs[k] = int(kwargs[k])
+            elif not isinstance(kwargs[k], (int, float)):
+                raise ValueError('%s: constant "%s": expected integer or floating value, got "%s".'
+                                 % (type(self).__name__, k, type(kwargs[k]).__name__))
+        super(EnumType, self).__init__(**kwargs)
+
+    def __repr__(self):
+        return '%s(%s)' % (type(self).__name__, ', '.join('%s:%s' % (k, self[k]) for k in sorted(self.keys())))
+
+    def __getattr__(self, key):
+        if key in self:
+            return self[key]
+        if key == 'ctype':
+            return self.ctype
```

Comment:
This will probably cause infinite recursion.  Just let it fall through the normal case below since that is a regular attribute.